### PR TITLE
Moving portsOrch to global

### PR DESCRIPTION
**Pull request by stcheng**
- Moving portsOrch to global
- Adding intfsOrch dependency check
  - Adding struct IntfsEntry
  - Adding increase/decrease reference count functions
- Removing some trailing '\n's in logs

### CODE REVIEW

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -61,24 +78,29 @@ void IntfsOrch::doTask(Consumer &consumer)
...
 
             if (!port.m_rif_id)
             {
                 addRouterIntfs(port);
-                m_intfs[alias] = IpAddresses();
+                IntfsEntry intfs_entry;
+                intfs_entry.ip_addresses = IpAddresses();
+                intfs_entry.ref_count = 0;
+                m_syncdIntfses[alias] = intfs_entry;
+
```
Comment (stcheng) [inline orchagent/intfsorch.cpp]: empty line.

File: orchagent/intfsorch.cpp [hunk 2]
```diff
@@ -174,20 +199,44 @@ void IntfsOrch::doTask(Consumer &consumer)
...
+                continue;
+            }
+
+            SWSS_LOG_NOTICE("Remove packet action trap route ip:%s\n", ip_prefix.getIp().to_string().c_str());
+            decreaseRouterIntfsRefCount(alias);
+            m_syncdIntfses[alias].ip_addresses.remove(ip_prefix.getIp());
+            it = consumer.m_toSync.erase(it);
+        }
+    }
+    cleanUpRouterInterfaces();
```
Comment (stcheng) [inline orchagent/intfsorch.cpp]: This needs to be moved into the DEL_COMMAND block and the task cannot be erased until the router interface is cleaned up.

File: orchagent/intfsorch.cpp [hunk 3]
```diff
@@ -174,20 +199,44 @@ void IntfsOrch::doTask(Consumer &consumer)
...
+}
+
+void IntfsOrch::cleanUpRouterInterfaces()
```
Comment (stcheng) [inline orchagent/intfsorch.cpp]: maybe no need to iterate all items checking the ref count.

**Comment by stcheng** [comment]
Adding port_status_change notification handling function.

File: orchagent/intfsorch.cpp [hunk 4]
```diff
@@ -15,18 +14,35 @@ extern sai_object_id_t gVirtualRouterId;
 extern sai_router_interface_api_t*  sai_router_intfs_api;
 extern sai_route_api_t*             sai_route_api;
 
-IntfsOrch::IntfsOrch(DBConnector *db, string tableName, PortsOrch *portsOrch) :
-        Orch(db, tableName), m_portsOrch(portsOrch)
+extern PortsOrch *gPortsOrch;
+
+IntfsOrch::IntfsOrch(DBConnector *db, string tableName) :
+        Orch(db, tableName)
+{
```
Comment (kcudnik) [inline orchagent/intfsorch.cpp]: SWSS_LOG_ENTER();

File: orchagent/intfsorch.cpp [hunk 5]
```diff
@@ -15,18 +14,35 @@ extern sai_object_id_t gVirtualRouterId;
...
+}
+
+sai_object_id_t IntfsOrch::getRouterIntfsId(string alias)
+{
+    Port port;
```
Comment (kcudnik) [inline orchagent/intfsorch.cpp]: SWSS_LOG_ENTER();

File: orchagent/neighorch.cpp [hunk 6]
```diff
@@ -7,19 +7,24 @@
 extern sai_neighbor_api_t*         sai_neighbor_api;
 extern sai_next_hop_api_t*         sai_next_hop_api;
 
+extern PortsOrch *gPortsOrch;
+
+NeighOrch::NeighOrch(DBConnector *db, string tableName, IntfsOrch *intfsOrch) :
+        Orch(db, tableName), m_intfsOrch(intfsOrch)
+{
```
Comment (kcudnik) [inline orchagent/neighorch.cpp]: SWSS_LOG_ENTER();

File: orchagent/notifications.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+sai_switch_notification_t switch_notifications
+{
+    NULL,
+    NULL,
+    on_port_state_change,
+    NULL,
+    NULL,
+    NULL
+};
```
Comment (kcudnik) [inline orchagent/notifications.cpp]: do you want to handle on switch shutdown ? in case syncd will crash ?

File: orchagent/orchdaemon.cpp [hunk 8]
```diff
@@ -7,20 +7,16 @@
...
+OrchDaemon::OrchDaemon(DBConnector *applDb) :
+        m_applDb(applDb)
 {
-    m_applDb = nullptr;
-    m_asicDb = nullptr;
 }
 
 OrchDaemon::~OrchDaemon()
 {
-    if (m_applDb)
```
Comment (kcudnik) [inline orchagent/orchdaemon.cpp]: SWSS_LOG_ENTER();

File: orchagent/orchdaemon.cpp [hunk 9]
```diff
@@ -7,20 +7,16 @@
...
 using namespace swss;
 
-OrchDaemon::OrchDaemon()
+/* Global variable gPortsOrch declared */
+PortsOrch *gPortsOrch;
+
+OrchDaemon::OrchDaemon(DBConnector *applDb) :
+        m_applDb(applDb)
 {
-    m_applDb = nullptr;
```
Comment (kcudnik) [inline orchagent/orchdaemon.cpp]: SWSS_LOG_ENTER();

**Comment by kcudnik** [comment]
Missing some function entry logs, commented on those which i found, not sure if all

File: orchagent/notifications.cpp (hunk 7 above)
Comment (stcheng) [inline orchagent/notifications.cpp]: yes. will be in another pull request.

**Comment by stcheng** [comment]
Updated. Add entry logs on necessary places.

File: orchagent/intfsorch.cpp [hunk 10]
```diff
@@ -250,26 +235,74 @@ bool IntfsOrch::addRouterIntfs(Port &port, sai_object_id_t virtual_router_id, Ma
...
+    status = sai_route_api->remove_route(&unicast_route_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to remove action trap route ip:%s %d\n", ip_prefix.getIp().to_string().c_str(), status);
+        return false;
+    }
+
+    SWSS_LOG_NOTICE("Remove packet action trap route ip:%s\n", ip_prefix.getIp().to_string().c_str());
+    decreaseRouterIntfsRefCount(alias);
+    m_syncdIntfses[alias].ip_addresses.remove(ip_prefix.getIp());
```
Comment (lguohan) [inline orchagent/intfsorch.cpp]: the counterpart, adding them is not in addRouterIntf. why deleting them should be in removeRouterIntf?

Comment (stcheng) [inline orchagent/intfsorch.cpp]: hmm good point.  I'll move them to the addRouterIntf function().

**Comment by lguohan** [comment]
> 

I feel this should be part of addRouterIntfs()

---
Refers to: orchagent/intfsorch.cpp:110 in 5ab3f5b. [](commit_id = 5ab3f5bcda3d3ea2cb7802506068521f1335e35d, deletion_comment = False)

**Comment by lguohan** [comment]
>             unicast_route_entry.destination.mask.ip4 = 0xFFFFFFFF;

i feel this should also be port of addRouterIntfs.

---
Refers to: orchagent/intfsorch.cpp:146 in 5ab3f5b. [](commit_id = 5ab3f5bcda3d3ea2cb7802506068521f1335e35d, deletion_comment = False)

File: orchagent/intfsorch.cpp [hunk 11]
```diff
@@ -15,18 +14,36 @@ extern sai_object_id_t gVirtualRouterId;
 extern sai_router_interface_api_t*  sai_router_intfs_api;
 extern sai_route_api_t*             sai_route_api;
 
-IntfsOrch::IntfsOrch(DBConnector *db, string tableName, PortsOrch *portsOrch) :
-        Orch(db, tableName), m_portsOrch(portsOrch)
+extern PortsOrch *gPortsOrch;
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: Move 'extern' to header file.

File: orchagent/intfsorch.cpp [hunk 12]
```diff
@@ -250,26 +235,74 @@ bool IntfsOrch::addRouterIntfs(Port &port, sai_object_id_t virtual_router_id, Ma
...
-bool IntfsOrch::removeRouterIntfs(Port &port)
+bool IntfsOrch::removeRouterIntfs(string alias, IpPrefix ip_prefix)
 {
     SWSS_LOG_ENTER();
 
-    sai_status_t status = sai_router_intfs_api->remove_router_interface(port.m_rif_id);
+    if (m_syncdIntfses.find(alias) == m_syncdIntfses.end())
+        return true;
+
+    if (m_syncdIntfses[alias].ref_count > 2)
```
Comment (lguohan) [inline orchagent/intfsorch.cpp]: >if (m_syncdIntfses[alias].ref_count > 2) [](start = 4, length = 40)

for subnet and interface address IP route, can we count them as internal ref_count?

File: orchagent/intfsorch.cpp [hunk 13]
```diff
@@ -15,18 +14,36 @@ extern sai_object_id_t gVirtualRouterId;
...
-        Orch(db, tableName), m_portsOrch(portsOrch)
+extern PortsOrch *gPortsOrch;
+
+IntfsOrch::IntfsOrch(DBConnector *db, string tableName) :
+        Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+}
+
+sai_object_id_t IntfsOrch::getRouterIntfsId(string alias)
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: const string&

File: orchagent/intfsorch.cpp (hunk 12 above)
Comment (stcheng) [inline orchagent/intfsorch.cpp]: what about just do not count them to the ref_count?

**Comment by lguohan** [comment]
>             unicast_route_entry.destination.addr_family = SAI_IP_ADDR_FAMILY_IPV4;

did we add ipv6 support yet?

---
Refers to: orchagent/intfsorch.cpp:144 in 5ab3f5b. [](commit_id = 5ab3f5bcda3d3ea2cb7802506068521f1335e35d, deletion_comment = False)

File: orchagent/intfsorch.cpp [hunk 14]
```diff
@@ -15,18 +14,36 @@ extern sai_object_id_t gVirtualRouterId;
...
+{
+    Port port;
+    assert(gPortsOrch->getPort(alias, port));
+    assert(port.m_rif_id);
+    return port.m_rif_id;
+}
+
+void IntfsOrch::increaseRouterIntfsRefCount(string alias)
+{
+    m_syncdIntfses[alias].ref_count++;
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: Is alias always in the container? Add assert or runtime check?

File: orchagent/intfsorch.cpp [hunk 15]
```diff
@@ -61,24 +78,28 @@ void IntfsOrch::doTask(Consumer &consumer)
...
+                it++;
                 continue;
             }
 
             if (!port.m_rif_id)
             {
                 addRouterIntfs(port);
-                m_intfs[alias] = IpAddresses();
+                IntfsEntry intfs_entry;
+                intfs_entry.ip_addresses = IpAddresses();
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: ip_addresses is already initialized by default constructor. Why constructing another one and assigning?

File: orchagent/intfsorch.cpp [hunk 16]
```diff
@@ -250,26 +235,74 @@ bool IntfsOrch::addRouterIntfs(Port &port, sai_object_id_t virtual_router_id, Ma
...
-    m_portsOrch->setPort(port.m_alias, port);
+    gPortsOrch->setPort(port.m_alias, port);
 
     SWSS_LOG_NOTICE("Create router interface for port %s", port.m_alias.c_str());
 
     return true;
 }
 
-bool IntfsOrch::removeRouterIntfs(Port &port)
+bool IntfsOrch::removeRouterIntfs(string alias, IpPrefix ip_prefix)
```
Comment (qiluo-msft) [inline orchagent/intfsorch.cpp]: const string& alias, const IpPrefix& ip_prefix

File: orchagent/intfsorch.cpp (hunk 11 above)
Comment (stcheng) [inline orchagent/intfsorch.cpp]: discuss this in future pull requests.

**Comment by stcheng** [comment]
Update the code.
